op-chain-ops/script/precompile.go

```diff
@@ -0,0 +1,447 @@
+		if err != nil {
+			return "", "", fmt.Errorf("unrecognized slice-elem type: %w", err)
+		}
+		if internalTyp != "" {
+			return "", "", fmt.Errorf("nested internal types not supported: %w", err)
+		}
+		return elemABITyp + "[]", "", nil
+	case reflect.Struct:
+		if typ.AssignableTo(abiInt256Type) {
+			return "int256", "", nil
+		}
+		if typ.ConvertibleTo(typeFor[big.Int]()) {
+			return "uint256", "", nil
+		}
+		// We can parse into abi.TupleTy in the future, if necessary
+		return "", "", fmt.Errorf("structs are not supported, cannot handle type %s", typ)
+	case reflect.Pointer:
+		elemABITyp, internalTyp, err := goTypeToSolidityType(typ.Elem())
+		if err != nil {
+			return "", "", fmt.Errorf("unrecognized pointer-elem type: %w", err)
+		}
+		return elemABITyp, internalTyp, nil
+	default:
+		return "", "", fmt.Errorf("unrecognized typ: %s", typ)
+	}
+}
+
+// setupFields registers all exported non-ignored fields as public ABI getters.
+// Fields and methods of embedded structs are registered along the way.
+func (p *Precompile[E]) setupFields(val *reflect.Value) error {
+	if val.Kind() == reflect.Pointer {
+		if val.IsNil() {
+			return fmt.Errorf("cannot setupFields of nil value (type: %s)", val.Type())
+		}
+		inner := val.Elem()
+		if err := p.setupFields(&inner); err != nil {
+			return fmt.Errorf("failed to setupFields of inner pointer type: %w", err)
+		}
+		return nil
+	}
+	if val.Kind() != reflect.Struct {
+		return nil // ignore non-struct types
+	}
+	typ := val.Type()
+	fieldCount := val.NumField()
+	for i := 0; i < fieldCount; i++ {
+		fieldTyp := typ.Field(i)
+		if !fieldTyp.IsExported() {
+			continue
+		}
+		if tag, ok := fieldTyp.Tag.Lookup("evm"); ok && tag == "-" {
+			continue
+		}
+		fieldVal := val.Field(i)
+		if fieldTyp.Anonymous {
+			// process methods and inner fields of embedded fields
+			if err := p.setupMethods(&fieldVal); err != nil {
+				return fmt.Errorf("failed to setup methods of embedded field %s (type: %s): %w",
+					fieldTyp.Name, fieldTyp.Type, err)
+			}
+			if err := p.setupFields(&fieldVal); err != nil {
+				return fmt.Errorf("failed to setup fields of embedded field %s (type %s): %w",
+					fieldTyp.Name, fieldTyp.Type, err)
+			}
+			continue
+		}
+		if err := p.setupStructField(&fieldTyp, &fieldVal); err != nil {
+			return fmt.Errorf("failed to setup struct field %s (type %s): %w", fieldTyp.Name, fieldTyp.Type, err)
+		}
+	}
+	return nil
+}
+
+// setupStructField registers a struct field as a public-getter ABI method.
+func (p *Precompile[E]) setupStructField(fieldDef *reflect.StructField, fieldVal *reflect.Value) error {
+	abiFunctionName := fieldDef.Name
+	if len(abiFunctionName) == 0 {
+		return fmt.Errorf("ABI name of %s must not be empty", fieldDef.Name)
+	}
+	if lo := strings.ToLower(abiFunctionName[:1]); lo != abiFunctionName[:1] {
+		abiFunctionName = lo + abiFunctionName[1:]
+	}
+
+	methodSig := abiFunctionName + "()"
+	byte4Sig := bytes4(methodSig)
```

Comment:
This seems a little fragile, are there any `fieldDef.Name` that wouldn't work?

Reply:
Added comments to clarify what is happening here. There is some validation of the field definition already, all field names would work (if the name is a sensible ABI signature is up to the user).